AwsCryptographyPrimitives/runtimes/python/src/aws_cryptography_primitives/internaldafny/extern/ECDH.py

```diff
@@ -0,0 +1,394 @@
+            try:
+                private_key = ec.generate_private_key(
+                    maybe_ecc_algorithm.value.value
+                )
+
+                private_key_pem_bytes = private_key.private_bytes(Encoding.PEM, PrivateFormat.PKCS8, NoEncryption())
+                public_key_der_bytes = private_key.public_key().public_bytes(Encoding.DER, PublicFormat.SubjectPublicKeyInfo)
+
+                return CreateExternEccKeyGenSuccess(EccKeyPair_EccKeyPair(
+                    publicKey=_dafny.Seq(public_key_der_bytes),
+                    privateKey=_dafny.Seq(private_key_pem_bytes)
+                ))
+            except Exception as e:
+                return CreateExternEccKeyGenFailure(
+                    _smithy_error_to_dafny_error(
+                        AwsCryptographicPrimitivesError(
+                            message=str(e)
+                        )
+                    )
+                )
+        else:
+            return CreateExternEccKeyGenFailure(
+                    _smithy_error_to_dafny_error(
+                        ValueError(
+                            "SM2 not supported."
+                        )
+                    )
+                )
+
+
+class ECCUtils:
+
+    CURVE_TO_ECC_SECRET_LENGTH_MAP = {
+        "secp256r1": int(256 / 8),
+        "secp384r1": int(384 / 8),
+        "secp521r1": int(521 / 8 + 1)
+    }
+    
+    def ParsePublicKey(dafny_publicKey: _dafny.Seq) -> Wrappers.Result:
+        try:
+            public_key_bytes = bytes(dafny_publicKey)
+            public_key = load_der_public_key(public_key_bytes)
+            public_bytes = public_key.public_bytes(Encoding.DER, PublicFormat.SubjectPublicKeyInfo)
+
+            return CreateExternParsePublicKeySuccess(
+                _dafny.Seq(
+                    public_bytes
+                )
+            )
+        except Exception as e:
+            return CreateExternParsePublicKeyError(
+                _smithy_error_to_dafny_error(
+                    AwsCryptographicPrimitivesError(
+                        message=str(e)
+                    )
+                )
+            )
+    
+    def GetPublicKey(dafny_eccAlgorithm, dafny_privateKey):
+        maybe_ecc_algorithm = ECCAlgorithms.eccAlgorithm(dafny_eccAlgorithm)
+        if maybe_ecc_algorithm.is_Failure:
+            return CreateExternGetPublicKeyFromPrivateError(
+                maybe_ecc_algorithm.error
+            )
+        
+        curve = maybe_ecc_algorithm.value
+
+        if curve.name != "SM2PKE": # magic string?
+            try:
+                # get order for parsed private key
+                private_key_pem = bytes(dafny_privateKey.pem)
+                private_key = load_pem_private_key(private_key_pem, None)
+                private_key_curve = curve_mapping[private_key.curve.name]
+                private_key_order = private_key_curve.order
+
+                # get expected curve order
+                ecdsa_curve = curve_mapping[curve.name]
```

Comment:
let's change this to a different name cause it is incorrect to refer to this as ecdsa since these keys are not for signing.

How about?
```suggestion
                ecdh_curve = curve_mapping[curve.name]
```